services/galley/src/Galley/Data.hs

```diff
@@ -691,17 +712,26 @@ privateRole = PrivateAccessRole
 privateOnly :: Set Access
 privateOnly = Set [PrivateAccess]
 
-toMappedOrLocalId :: (Id (Opaque a), Maybe (Id (Remote a)), Maybe Domain) -> MappedOrLocalId a
+type MappedOrLocalIdRow a = (Id (Opaque a), Maybe (Id (Remote a)), Maybe Domain)
+
+toMappedOrLocalId :: (Log.MonadLogger m, MonadThrow m) => MappedOrLocalIdRow a -> m (MappedOrLocalId a)
 toMappedOrLocalId = \case
-  (Id mappedId, Just remoteId, Just domain) ->
-    Mapped (IdMapping (Id mappedId) (Qualified remoteId domain))
-  (Id localId, Nothing, Nothing) ->
-    Local (Id localId)
-  (Id localId, _, _) ->
-    -- TODO(mheinzel): return either, raise error
-    Local (Id localId)
-
-fromMappedOrLocalId :: MappedOrLocalId a -> (Id (Opaque a), Maybe (Id (Remote a)), Maybe Domain)
+  (mappedId, Just remoteId, Just domain) ->
+    pure $ Mapped (IdMapping (coerce mappedId) (Qualified remoteId domain))
+  (localId, Nothing, Nothing) ->
+    pure $ Local (coerce localId)
+  invalid -> do
+    -- This should never happen as we always write rows with either both or none of these
+    -- values.
+    -- FUTUREWORK: we could try to recover from this situation by checking if an ID mapping
+    -- for this mapped ID exists (and potentially even repair the row). At the moment, the
+    -- problem seems unlikely enough not to warrant the complexity, though.
+    -- In some cases it could also be better not to fail, but skip this entry, e.g. when
+    -- deleting a user, we should remove him from all conversations we can, not stop halfway.
+    Log.err . Log.msg $ "Invalid remote ID in database row: " <> show invalid
+    throwM internalError
```

Comment:
Maybe we should put this in the error message instead of the log

Reply:
I did both now.